### PR TITLE
Add PackageIdentity type

### PR DESCRIPTION
**Pull request by mattt**
### Motivation:

See #3032

### Modifications:

- Replaced `PackageReference.PackageIdentity` with top-level `PackageIdentity` type
- Replaced existing calls to `PackageReference.computeIdentity` to initialization of new `PackageIdentity` type.
- Changed access level of `PackageReference.computeDefaultName` to `internal`, limiting its use outside of `PackageModel` module.
- Extracted `testPackageReference` into new, separate test case
- Updated package identity to be configurable via top-level `_useLegacyIdentities` variable. This acts as a setter for a `static internal` property on `PackageIdentity` that configures how package identity is calculated via dependency injection.   

### Result:

This PR shouldn't affect current behavior. Rather, it serves to codify existing determinations of package identity and to tee up subsequent PRs to accommodate the new package identity semantics required for package registry support.

### CODE REVIEW

File: Sources/PackageModel/PackageIdentity.swift
Comment (mattt) [inline Sources/PackageModel/PackageIdentity.swift:16]: Aside: I wish `@warn_unqualified_access` was available for top-level variable declarations (not just functions).

**Comment by mattt** [comment]
@swift-ci please smoke test

File: Sources/PackageModel/PackageIdentity.swift
Comment (neonichu) [inline Sources/PackageModel/PackageIdentity.swift:26]: I think the identity provider also has to provide equality for package identities, not just an initializer. For the "real" new implementation of identity we will need a staggered comparison that's not just comparing strings.

**Review by neonichu — APPROVED** [review]
Thanks, this looks great! 

I would just make the identity provider a bit more powerful as mentioned.

File: Sources/PackageModel/PackageIdentity.swift
Comment (abertelrud) [inline Sources/PackageModel/PackageIdentity.swift:101]: Should this description also mention path-based external dependencies?  They do not use `file://`, which is only used when the local path contains an actual local Git repository and not just an unversioned package.  Or is package identity only a thing for actual remote packages (not root or local)?

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [inline Sources/PackageModel/PackageReference.swift:31]: This will break clients that show a package name before the package has been loaded.  What is the new API for them to use?

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [inline Sources/PackageModel/PackageReference.swift:33]: This will break clients that currently maintain mappings from packages to their own objects, such as for IDEs.  What is the new API they should use?  Should they just call `PackageIdentity(url: url)` instead?

Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:33]: What about for local paths (specified by file path) and roots?  How are their identities formed in the new API?

Comment (tomerd) [reply Sources/PackageModel/PackageReference.swift:33]: could PackageReference expose the identity it is initialized with and such clients can work off of that?

File: Sources/PackageModel/PackageReference.swift
Comment (tomerd) [reply Sources/PackageModel/PackageReference.swift:31]: could PackageReference expose the identity it is initialized with and such clients can work off of that?

**Review by tomerd — APPROVED** [review]
this looks great, thanks @mattt 

there are two question @abertelrud brought up that should be answered but I hope they are straight forward by exposing the identity on the ref

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:31]: This is about being able to come up with a default name to show before the package has been loaded.  We could show the raw identity string, but it would be good to have a centralized place (of some form) to show the "display name" to use for a particular identity.

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:33]: I think the right thing here is probably for clients to create a `PackageIdentity` from the given URL.  The question about local paths was more of a question about whether the diffs in this PR would handle that properly.  The URL that is passed to `PackageIdentity` would need to support local paths as well as `file://` URLs (which are different, the former being a local unversioned package reference and the latter being a local Git repository).

**Comment by abertelrud** [comment]
Right, I don't think they are significant issues (unless it turns out that local packages are no longer supported, but I don't think that's the case if the unit tests pass) but since clients will need to adapt, I'm trying to identify what the downstream changes will need to be.  From the unit tests it looks as if it can mostly be handled by instantiating a `PackageIdentity` from the URL strings.

File: Sources/PackageModel/PackageReference.swift
Comment (neonichu) [reply Sources/PackageModel/PackageReference.swift:33]: Yep, agreed, supporting these cases fully is what I'll be doing in a follow-up. Do you think the interface needs to change to allow this? I would think we can pass in a file URL and depending on whether there's a repo at that location or not, we would get different behavior in terms of what we use for the identity (a repo could use the fingerprint we talked about, if it is just a directory we would essentially fallback to last path component).

**Review by abertelrud — APPROVED** [review]

File: Sources/PackageModel/PackageReference.swift
Comment (neonichu) [reply Sources/PackageModel/PackageReference.swift:31]: I'd say we should keep this `public` in this case and figure this out separately as it is not strictly needed for Mattt to make progress.

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:33]: I think that the API should ideally allow creation of a package identity from either an `AbsolutePath` (for a local unversioned path reference) or a `URL` (which designates either a local repository via `file://` or a remote repository using some other scheme).  What the implementation does under-the-hood is none of the client's business, but that would allow the client to express the two quite fundamentally different kinds of package reference (specifically one is versioned and the other is taken as given).

For now we can have a single string that either has or doesn't have a `file` scheme but ideally there would be a clearer API.  But that can be added later, as you suggest (I'm trying to not keep requesting changes from Mattt here).

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:31]: That would make it easier, but in the short term we can also switch clients to using `PackageIdentity.description` and add back a minimal API later.

File: Sources/PackageModel/PackageIdentity.swift
Comment (mattt) [reply Sources/PackageModel/PackageIdentity.swift:101]: That's what I was trying to get at — my conceit here is that a file path is _actually_ a `file:///` URL with an implicit scheme. Any package dependency, local or remote, is described by a URL: local use `file:///` URLs and remote use any other scheme. `CanonicalPackageIdentity` distinguishes between local from remote by preserving a leading slash (for example, you couldn't shadow a remote URL by having a root directory named `/github.com/`). Does that all make sense?

File: Sources/PackageModel/PackageReference.swift
Comment (mattt) [reply Sources/PackageModel/PackageReference.swift:31]:
```suggestion
    public static func computeDefaultName(fromURL url: String) -> String {
```

I agree with @neonichu's take. How we decide to migrate this particular API is tertiary and severable from the rest of this PR. If we want to change the public API (and I think that we should), we can always do that later.

File: Sources/PackageModel/PackageIdentity.swift
Comment (mattt) [reply Sources/PackageModel/PackageIdentity.swift:26]: If we require `PackageIdentityProvider` to be `Equatable`, then it can only be used as a generic constraint, because it has `Self` requirements in `==`.

I found a way to work around this, but it's a good bit more complex than if we forego equatability:

```swift
public var _useLegacyIdentities: Bool = true {
    willSet {
        PackageIdentity.provider = newValue ? LegacyPackageIdentity.init : CanonicalPackageIdentity.init
    }
}

private typealias PackageIdentityProvider = (String) -> PackageIdentityProtocol

private protocol PackageIdentityProtocol: LosslessStringConvertible {
    init(_ string: String)
}

/// The canonical identifier for a package, based on its source location.
public struct PackageIdentity: LosslessStringConvertible, Hashable {
    fileprivate static var provider: PackageIdentityProvider = LegacyPackageIdentity.init
    public let description: String
    public init(_ string: String) {
        self.description = Self.provider(string).description
    }
}
```

Any strong preference between the two?

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageModel/PackageIdentity.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageIdentity.swift:101]: Understood, but I think that's too subtle and confusing of a way to represent it.  And checking the file system for whether a particular path has a .git isn't a good way to do it either.  I think there should be a clear way to represent that a package is either an unversioned take-it-as-given path in the file system or a repository that just happens to be in the local file system.  Historically there has been value in using no scheme for the former and file: for the latter.  I think it would be a mistake to conflate them, even it's through an extra slash character.

**Comment by mattt** [comment]
@abertelrud  @tomerd  @neonichu  Thanks for reviewing my changes. I just pushed 766ea50f, which is responsive to [this discussion thread](https://github.com/apple/swift-package-manager/pull/3057#pullrequestreview-531939671). This commit takes the additional step to distinguish between `PackageIdentity` values constructed from absolute paths and URLs ([per @abertelrud 's comment](https://github.com/apple/swift-package-manager/pull/3057#discussion_r524924009)). Right now, the `url` version takes a `String`, but that could be migrated to `Foundation.URL` or some other URL type in the future.

Does this seem like the right change for this PR?

File: Sources/PackageModel/PackageReference.swift
Comment (mattt) [reply Sources/PackageModel/PackageReference.swift:33]: 766ea50 creates two initializers `PackageIdentity(url: String)` and `PackageIdentity(path: AbsolutePath)` and goes through each existing call site to figure out which one should be used. (Most of the hassle came around tests, which had used hardcoded string literals.)

See https://github.com/apple/swift-package-manager/pull/3057#issuecomment-729131108.

**Comment by mattt** [comment]
@swift-ci please smoke test

File: Sources/PackageModel/PackageReference.swift
Comment (abertelrud) [reply Sources/PackageModel/PackageReference.swift:33]: Great, thank you!  That sounds like the right change from a client's perspective.

**Comment by abertelrud** [comment]
Thanks, @mattt!  I'll take a look at the new diffs right away, but that sounds like the right API from a client's perspective.

**Review by abertelrud — APPROVED** [review]
Thanks @mattt, this looks good to me know.  I realize that separate URL and AbsolutePath initializers for package identity are maybe tedious, but I think it will really help clients keep things straight.  In one client I worked on, this confusion, where everything was just a string, was a source of at least one bug.

**Comment by abertelrud** [comment]
Thanks again for these diffs, @mattt.  I'm adapting some of the downstream dependencies to it and will do some testing to avoid regressions.  I plan to wrap that up by tonight.  Are you OK with holding off until tomorrow to merge?

File: Sources/PackageModel/PackageIdentity.swift
Comment (neonichu) [reply Sources/PackageModel/PackageIdentity.swift]: I would just add a comparison function that's unrelated to `Equatable`. I think what we want to be able to do is compare two `PackageIdentity` instances using a comparison operation vended by `PackageIdentityProvider`.

**Comment by mattt** [comment]
Absolutely. Thanks for taking a look at the downstream consumers. Let me know if anything comes up in your review.

**Comment by abertelrud** [comment]
Great, thanks @mattt.  The changes that are technically needed to keep compiling turn out to be fairly minimal, and similar to unit test changes in this PR... the later changes to adapt to the change in package identity will be more interesting (but of course very useful!).  But that's for after this PR lands.

**Comment by abertelrud** [comment]
@mattt this seems to work fine for us downstream, after adapting some of the client code.  All ready to merge as far as you're concerned?

**Comment by mattt** [comment]
@abertelrud I just gave everything another once-over, and I think this is ready to merge. Feel free to do that whenever you're ready.

I'd like to continue our conversation from https://github.com/apple/swift-package-manager/pull/3057/files#r525394175, but that's not a blocker here.

**Comment by abertelrud** [comment]
Great, thanks! Merging.

**Comment by tomerd** [comment]
🥳 🎊

**Pull request merged**
